python/amici/sbml_preprocessing.py

```diff
@@ -0,0 +1,106 @@
+import libsbml
+
+
+def flatten_sbml(sbml: libsbml.Model):
+    """Reformulate parts of the SBML, in order to allow
+        import of some SBML features.
+
+    Arguments:
+        sbml: Instance of libsbml.Model`.
+
+    Returns:
+
+    Raises:
+    """
+
+    process_rate_rules(sbml)
+    process_species_assignment_rules(sbml)
+
+
+def process_rate_rules(sbml: libsbml.Model):
+    """ Reformulate rate rules by an reaction of the form
+     R: -> X with rate f(x).
+
+    Arguments:
+        sbml: Instance of libsbml.Model`.
+
+    Returns:
+
+    Raises:
+    """
+
+    # loop backwards over rules, so that one can delete rules inside the loop!
+    for i in range(sbml.num_rules - 1, -1, -1):
+
+        rule = sbml.getRule(i)
+        rule_variable = sbml.getElementBySId(rule.getVariable())
+
+        if rule.isRate() and isinstance(rule_variable, libsbml.Species):
+
+            # create dummy reaction R: -> X with rate f(X)
+            reaction = sbml.createReaction()
+            reaction.setId('d_dt_' + rule_variable.getId())
+            reaction.setName('d_dt_' + rule_variable.getName())
+            reaction.setReversible(False)
+            reaction.setFast(False)
+
+            product = reaction.createProduct()
+            product.setSpecies(rule.getVariable())
+            product.setConstant(False)
+
+            k = reaction.createKineticLaw()
+            k.setMath(rule.getMath())
+
+            # add modifiers
+            for species in sbml.getListOfSpecies():
+
+                if species.getId() in k.getFormula():
+                    reaction.addModifier(
+                        sbml.getElementBySId(species.getId())
+                    )
+            # remove rule
+            rule.removeFromParentAndDelete()
+
+
+def process_species_assignment_rules(sbml: libsbml.Model):
+    """ Reformulate species assignment rules by replacing the species by
+    a parameter.
```

Comment:
Wait, I think that won't really work. The concentration of the assigned-to species is supposed to be able to change over time. Parameters don't. Wouldn't we have to replace all occurrences of the assigned-to species by the right hand side of the assignment?